autolab/Autolab
# Handle malformed scoreboard results from autograder, fix error handling for scoreboards

## Summary

### Summary generated by Reviewpad on 01 Nov 23 04:11 UTC
This pull request includes the following changes:

- Added `.error-header` class to `style.css.scss`
- Refactored error handling in `scoreboards_controller.rb`
- Added `_error_icon.html.erb` partial
- Updated `_form.html.erb` and `show.html.erb` to handle error rendering


## Description

- Detect autograded scoreboard results that are not formatted as an array during entry creation stage, return error flash to instructors when this occurs
- Fix error handling for `ScoreboardsController#show` such that it properly renders the scoreboard edit page with the error details at the top of the page

## Motivation and Context



CMU Prod recently faced an issue where the scoreboard for Attacklab was breaking with the following error, when viewed as a student:
```
Subject: [CMU-Autolab-Prod Error] scoreboards#show (ActionView::Template::Error) "undefined method `[]' for 95.0:Float"
An ActionView::Template::Error occurred in scoreboards#show:

  undefined method `[]' for 95.0:Float
  app/views/scoreboards/show.html.erb:64
```

This happened because the scoreboard rendering code expected each scoreboard entry to have an array of scores to display, but the assessment's autograder was returning a single floating point number. In addition, when viewing the scoreboard as an instructor, another error occured:

```
Subject: [CMU-Autolab-Prod Error] scoreboards#show (ActionView::MissingTemplate) "Missing partial courses/_course with {:locale=>...
An ActionView::MissingTemplate occurred in scoreboards#show:

  Missing partial courses/_course with {:locale=>[:en], :formats=>[:html], :variants=>[], :handlers=>[:raw, :erb, :html, :builder, :ruby, :coffee, :jbuilder]}. Searched in:
  * "/opt/autolab/app/views"
  * "/usr/local/lib/ruby/gems/2.7.0/gems/doorkeeper-5.6.6/app/views"
  * "/usr/local/lib/ruby/gems/2.7.0/gems/devise-4.8.1/app/views"
  * "/usr/local/lib/ruby/gems/2.7.0/gems/fomantic-ui-sass-2.8.8/app/views"
  * "/usr/local/lib/ruby/gems/2.7.0/gems/actiontext-6.0.5/app/views"
  * "/usr/local/lib/ruby/gems/2.7.0/gems/actionmailbox-6.0.5/app/views"

  app/controllers/scoreboards_controller.rb:118:in `rescue in block in show'
```

This error occured as there is no `_course` partial, but the scoreboards controller was attempting to render such a partial when facing the previous scoreboard error.

These fixes should make the scoreboard no longer fail to render when the autograder results are malformed, and provide helpful error information to instructors when results are malformed.

## How Has This Been Tested?



- Use the following assessment (it includes an autograder which returns malformed scoreboard results)
[randomlab2.tar.zip](https://github.com/autolab/Autolab/files/12664885/randomlab2.tar.zip)
- Upload submission(s) to the assessment as a student
- View the scoreboard as a student, note that the scoreboard now has error icons where scores usually should exist
  <img width="1512" alt="Screenshot 2023-09-19 at 1 36 29 PM" src="https://github.com/autolab/Autolab/assets/25730111/8fc92648-40a9-4b33-b45a-9bbd528e2e06">
- View the scoreboard as an instructor, note the error flash at the top
  <img width="1512" alt="Screenshot 2023-09-19 at 1 20 27 PM" src="https://github.com/autolab/Autolab/assets/25730111/bca812df-48ee-430c-b71e-f61b5cd5ead0">
- Add `raise StandardError` between lines 81-82 in `app/controllers/scoreboards_controller.rb`
- As an instructor, view the scoreboard again, see the following page appears:
  <img width="1512" alt="Screenshot 2023-09-19 at 1 20 14 PM" src="https://github.com/autolab/Autolab/assets/25730111/530ff360-7287-4d02-9d2e-7599278cf622">
- verify normal scoreboards render fine


## Types of changes


## Checklist:
